add short term (rolling) fourier transform primitives

ozzieD opened the pull request:
Resolves: https://github.com/alteryx/featuretools-tsfresh-primitives/issues/99

gsheni commented on featuretools_tsfresh_primitives/primitives/fft_aggregated.py:
```diff
@@ -30,3 +35,112 @@ def function(x):
         return function
+
+
+class ShortTermFftAggregated(TransformPrimitive):
```
Curious, why is this called short term?

ozzieD replied:
It's the most widely used term. I don't see any literature that refers to it as a "rolling" Fourier. 
https://stackoverflow.com/questions/20771080/is-a-rolling-fft-possible-and-could-it-be-of-use
https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.stft.html
https://en.wikipedia.org/wiki/Short-time_Fourier_transform

ozzieD commented:
Removed 3.7 support since checks were failing (we removed 3.7 support for feature tools). Also tsfresh still has issues with 3.10 so that has not been added.

thehomebrewnerd commented:
> Removed 3.7 support since checks were failing (we removed 3.7 support for feature tools). Also tsfresh still has issues with 3.10 so that has not been added.

You should also update `setup.py` to remove the two references to 3.7 in the `classifiers` section and `python_requires`.

thehomebrewnerd commented:
This is perhaps a bigger discussion, but with this PR, it seems we are changing the scope of this repo. Previously it was limited (I think) to only allowing tsfresh primitives to be used with Featuretools. Now, with these new primitives, we have created additional time series primitives beyond those from tsfresh. Just wondering if we should consider eventually renaming this to be more indicative of what is included.

Also, would it make sense to separate our the tests for our new primitives from the existing tsfresh tests in separate `test_*.py` files to have a more clear division between our custom primitives and the tsfresh primitives?

ozzieD commented:
> This is perhaps a bigger discussion, but with this PR, it seems we are changing the scope of this repo. Previously it was limited (I think) to only allowing tsfresh primitives to be used with Featuretools. Now, with these new primitives, we have created additional time series primitives beyond those from tsfresh. Just wondering if we should consider eventually renaming this to be more indicative of what is included.
> 
> Also, would it make sense to separate our the tests for our new primitives from the existing tsfresh tests in separate `test_*.py` files to have a more clear division between our custom primitives and the tsfresh primitives?

I can split it into a different test file. It's tough because these primitives are using the tsfresh functions. If we add them into featuretools we'll have a circular import problem. With this spike that I'm taking, I'm going to try and figure out a way we can just wrap any aggregate primitive into rolling transform primitive without having to define separate primitives for each.

thehomebrewnerd commented on featuretools_tsfresh_primitives/primitives/fft_aggregated.py:
```diff
@@ -30,3 +35,115 @@ def function(x):
+    ]
+    return_type = ColumnSchema(logical_type=Double, semantic_tags={"numeric"})
+
+    def __init__(self, aggtype, window_length=3, gap=0, min_periods=0):
```
Do you think we should provide a default for `aggtype`? I think we generally (always?) provide default values for primitive arguments.

ozzieD replied:
hmm I actually didn't notice this until now. I'll update these primitives as well as the orginal ones

thehomebrewnerd commented on featuretools_tsfresh_primitives/primitives/fft_aggregated.py:
```diff
@@ -30,3 +35,115 @@ def function(x):
+
+    Description:
+        Given a list of numbers and a corresponding list of
+        datetimes, return the spectral centroid (mean), variance, skew, and kurtosis of
```
Another use of `and` that might be better as `or`?

ozzieD replied:
Makes sense, I'll update this in the original primitive as well.

thehomebrewnerd replied:
Minor nitpick, but L46 still says `and` instead of `or`.

ozzieD replied:
my bad. resolved

thehomebrewnerd commented on featuretools_tsfresh_primitives/primitives/fft_coefficient.py:
```diff
@@ -40,3 +45,115 @@ def function(x):
+
+
+class ShortTermFftCoefficient(TransformPrimitive):
+    """Calculates the spectral centroid (mean), variance, skew, and kurtosis of
+    the absolute fourier transform spectrum of entries over a given window.
+
+    Description:
+        Given a list of numbers and a corresponding list of
+        datetimes, return the spectral centroid (mean), variance, skew, and kurtosis of
+        the absolute fourier transform spectrum of
+        the numeric values, starting at the row `gap` rows away from the current row and
+        looking backward over the specified time window
+        (by `window_length` and `gap`).
```
This docstring needs to be updated as it still references the values calculated by `ShortTermFftAggregated`

thehomebrewnerd commented on featuretools_tsfresh_primitives/primitives/fft_coefficient.py:
```diff
@@ -40,3 +45,115 @@ def function(x):
+        efficient.
+
+    Docstring source:
+    https://tsfresh.readthedocs.io/en/latest/api/tsfresh.feature_extraction.html#tsfresh.feature_extraction.feature_calculators.fft_aggregated
```
Update source to link to coefficient docs

thehomebrewnerd commented on featuretools_tsfresh_primitives/primitives/fft_coefficient.py:
```diff
@@ -40,3 +45,115 @@ def function(x):
+    ]
+    return_type = ColumnSchema(logical_type=Double, semantic_tags={"numeric"})
+
+    def __init__(self, coeff, attr, window_length=3, gap=0, min_periods=0):
```
Similar to my other comment - should we have default values for `coeff` and `attr` here?

ozzieD commented on featuretools_tsfresh_primitives/primitives/__init__.py:
```diff
@@ -70,4 +69,9 @@
 from .variance_larger_than_standard_deviation import VarianceLargerThanStandardDeviation
 
-SUPPORTED_PRIMITIVES = [value for value in locals().values() if isinstance(value, type)]
+SUPPORTED_PRIMITIVES = [
```
Yeah I'm not the biggest fan of how this was set up. This is really only used for `test_primitives.py` which doesn't accommodate our new transform primitive at all. I'll change the name.

thehomebrewnerd commented on featuretools_tsfresh_primitives/primitives/fft_coefficient.py:
```diff
@@ -40,3 +45,123 @@ def function(x):
+        \\ldots , n-1.
+
+    The resulting coefficients will be complex, this feature calculator can
+    return the real part (attr=="real"), the imaginary part (attr=="imag), the
+    absolute value (attr=""abs) and the angle in degrees (attr=="angle).
```
The quotes on the attr examples here need to be cleaned up.

ozzieD replied:
Looks like they need to be changed in the original as well. Good catch

thehomebrewnerd commented on featuretools_tsfresh_primitives/primitives/fft_coefficient.py:
```diff
@@ -40,3 +45,123 @@ def function(x):
+    return the real part (attr=="real"), the imaginary part (attr=="imag), the
+    absolute value (attr=""abs) and the angle in degrees (attr=="angle).
+
+    Description:
+        Given a list of numbers and a corresponding list of
+        datetimes, return the spectral centroid (mean), variance, skew, and kurtosis of
+        the absolute fourier transform spectrum of
+        the numeric values, starting at the row `gap` rows away from the current row and
+        looking backward over the specified time window
+        (by `window_length` and `gap`).
```
This seems to be the description for `ShortTermFftAggregated`

thehomebrewnerd reviewed (commented):
Just a couple more minor nitpicky docstring items, but not seeing anything else major to fix.

thehomebrewnerd commented on featuretools_tsfresh_primitives/primitives/fft_aggregated.py:
```diff
@@ -30,3 +35,115 @@ def function(x):
+
+    def appliable_fft_aggregated(self, series):
+        param = [{"aggtype": self.aggtype}]
+        return list(fft_aggregated(series.values, param=param))[0][1]
```
Last question - the `fft_aggregated` documentation (also the `fft_coefficient docs) say that the function returns a `pd.Series`. I'm a little unclear on what is in the series, and why we have to convert to a list and then select list items like this.

Unfortunately I can't seem to get this repo to install on my M1 Mac, so I'm not able to load and test on my own to understand this...

ozzieD replied:
after doing some poking around it looks like it was written like this because it returns a zip object

thehomebrewnerd replied:
I see...
